sdk/go/common/util/env/env.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A small library for creating consistent and documented environmental variable accesses.
+//
+// Public environmental variables should be declared as a module level variable.
+package env
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+)
+
+var Experimental = Bool("EXPERIMENTAL", "Enable experimental options and commands")
+
+type Var struct {
+	name        string
+	Value       Value
+	Description string
+	options     options
+}
+
+const Prefix = "PULUMI"
+
+func (v Var) Name() string {
+	return v.options.name(v.name)
+}
+
+func (v Var) Requires() []BoolValue { return v.options.prerequs }
+
+var envVars []Var
+
+// Variables is a list of variables declared.
+func Variables() []Var {
+	return envVars
+}
+
+type Option func(*options)
+
+type options struct {
+	prerequs []BoolValue
+	noPrefix bool
+}
+
+func (o options) name(underlying string) string {
+	if o.noPrefix {
+		return underlying
+	}
+	return Prefix + "_" + underlying
+}
+
+func Needs(val BoolValue) Option {
```

Comment:
Is there no version of this function with a `StringValue` argument?

Reply:
No, for two reasons:
1. I'm not aware of any place where we condition a set of variables on the result of a string value.
2. It's not clear what `Needs(MyString)` would mean. `BoolValue` has a clear truthy coercion. `FOO=1` and `FOO=true` is truthy, `FOO=0` or `unset FOO` is falsey. There isn't a clear surjection from `string` to `bool`. 

If there is a use case where that is necessary, we can add anything way to encode that functionality that puts the onus on the caller to define what true means. The first design I though of was a `When(func() bool)` style api. `Needs(v)` would then be `When(v.Value)`.